porch/controllers/packagevariantsets/pkg/controllers/packagevariantset/packagevariantset_controller.go

```diff
@@ -90,349 +90,159 @@ func (r *PackageVariantSetReconciler) Reconcile(ctx context.Context, req ctrl.Re
 	}()
 
 	if errs := validatePackageVariantSet(pvs); len(errs) > 0 {
-		validationErr := combineErrors(errs)
-		meta.SetStatusCondition(&pvs.Status.Conditions, metav1.Condition{
-			Type:    "Valid",
-			Status:  "False",
-			Reason:  "Invalid",
-			Message: validationErr,
-		})
-		return ctrl.Result{}, fmt.Errorf(validationErr)
-	} else {
-		meta.SetStatusCondition(&pvs.Status.Conditions, metav1.Condition{
-			Type:    "Valid",
-			Status:  "True",
-			Reason:  "Valid",
-			Message: "all validation checks passed",
-		})
+		setStalledConditionsToTrue(pvs, "ValidationError", combineErrors(errs))
+		return ctrl.Result{}, nil
 	}
 
-	upstream, err := r.getUpstreamPR(pvs.Spec.Upstream)
+	upstreamPR, err := r.getUpstreamPR(pvs.Spec.Upstream, prList)
 	if err != nil {
+		setStalledConditionsToTrue(pvs, "ValidationError", err.Error())
 		return ctrl.Result{}, err
```

Comment:
Should we return the error here? I would expect that if we set the stalled condition, we don't want to trigger an immediate new reconcile.

Reply:
When we return the error, that means reconcile again with exponential back off, is that correct?

So, in this case, we may just be waiting for the upstream to appear. So, we are stalled in the sense that some user action is needed; but it's not user action against *this* resource.

On other hand, it looks like right now we trigger a complete reconciliation of all PVSs in a namespace whenever there is a change to any PackageRevision. So, we should get triggered without any need for re-queueing, just by the nature of watching PackageRevisions. I wasn't sure though if there could be race conditions that this might resolve.

WDTY?

Reply:
I made this change and add an explanatory note.

Reply:
It seems like triggering reconciliation of all PVSs whenever any PackageRevision has a change seems unnecessary, but probably not a big concern until we have some scale. We do need to have watch on the PackageRevision resource, but we can probably be more selective of which PVSs gets reconciled for each PackageRevision.

Reply:
Yeah, I concur with that.